// appcenter-cli/dist/commands/codepush/lib/release-strategy/appcenter-release.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : new P(function (resolve) { resolve(result.value); }).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
const apis_1 = require("../../../../util/apis");
const fs = require("fs");
class AppCenterCodePushRelease {
    release(client, app, deploymentName, updateContentsZipPath, updateMetadata, token, serverUrl) {
        return __awaiter(this, void 0, void 0, function* () {
            yield apis_1.clientRequest((cb) => client.codePushDeploymentReleases.create(deploymentName, updateMetadata.appVersion, app.ownerName, app.appName, {
                packageProperty: fs.createReadStream(updateContentsZipPath),
                deploymentName1: deploymentName,
                description: updateMetadata.description,
                disabled: updateMetadata.isDisabled,
                mandatory: updateMetadata.isMandatory,
                noDuplicateReleaseError: false,
                rollout: updateMetadata.rollout,
            }, cb));
        });
    }
}
exports.default = AppCenterCodePushRelease;
