// appcenter-cli/dist/commands/test/lib/nunit-xml-util.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : new P(function (resolve) { resolve(result.value); }).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
const pfs = require("../../../util/misc/promisfied-fs");
const xml_util_1 = require("./xml-util");
const fs = require("fs");
const xmldom_1 = require("xmldom");
class NUnitXmlUtil extends xml_util_1.XmlUtil {
    mergeXmlResults(pathToArchive) {
        return __awaiter(this, void 0, void 0, function* () {
            const tempPath = yield pfs.mkTempDir("appcenter-uitestreports");
            let mainXml = null;
            const self = this;
            return this.getMergeXmlResultsPromise(pathToArchive, tempPath, (fullPath, entry) => {
                const xml = new xmldom_1.DOMParser(xml_util_1.XmlUtil.DOMParserConfig).parseFromString(fs.readFileSync(fullPath, "utf-8"), "text/xml");
                let name = "unknown";
                const matches = entry.path.match("^(.*)[_-]nunit[_-]report");
                if (matches && matches.length > 1) {
                    name = matches[1].replace(/\./gi, "_");
                }
                self.appendToTestNameTransformation(xml, `_${name}`);
                self.removeIgnoredTransformation(xml);
                self.removeEmptySuitesTransformation(xml);
                if (mainXml) {
                    mainXml = self.combine(mainXml, xml);
                }
                else {
                    mainXml = xml;
                }
            }, (resolve) => {
                resolve(mainXml);
            });
        });
    }
    getArchiveName() {
        return "nunit_xml_zip.zip";
    }
    combine(xml1, xml2) {
        this.combineTestResultsAttribute(xml1, xml2, "total");
        this.combineTestResultsAttribute(xml1, xml2, "errors");
        this.combineTestResultsAttribute(xml1, xml2, "failures");
        this.combineTestResultsAttribute(xml1, xml2, "not-run");
        this.combineTestResultsAttribute(xml1, xml2, "inconclusive");
        this.combineTestResultsAttribute(xml1, xml2, "ignored");
        this.combineTestResultsAttribute(xml1, xml2, "skipped");
        this.combineTestResultsAttribute(xml1, xml2, "invalid");
        const testSuitesParent = this.collectAllElements(xml1.documentElement, "test-results")[0];
        const testSuites = this.collectChildren(xml2.documentElement, "test-suite");
        testSuites.forEach((child) => {
            testSuitesParent.appendChild(child);
        });
        return xml1;
    }
    appendToTestNameTransformation(xml, text) {
        const testCases = this.collectAllElements(xml.documentElement, "test-case");
        testCases.forEach((testCase) => {
            const name = testCase.attributes.getNamedItem("name");
            if (name) {
                name.value = `${name.value}${text}`;
            }
        });
    }
    removeIgnoredTransformation(xml) {
        const testResults = this.collectAllElements(xml.documentElement, "test-results");
        testResults.forEach((testResult) => {
            const ignoredAttr = testResult.attributes.getNamedItem("ignored");
            if (ignoredAttr) {
                const notRunAttr = testResult.attributes.getNamedItem("not-run");
                if (notRunAttr) {
                    const notRun = Number(notRunAttr.value);
                    const ignored = Number(ignoredAttr.value);
                    notRunAttr.value = String(notRun - ignored);
                }
                ignoredAttr.value = "0";
            }
        });
        const elements = this.collectAllElements(xml.documentElement, "test-case");
        elements.forEach((element) => {
            const resultAttr = element.attributes.getNamedItem("result");
            if (resultAttr && resultAttr.value === "Ignored") {
                element.parentNode.removeChild(element);
            }
        });
    }
    removeEmptySuitesTransformation(xml) {
        const elements = this.collectAllElements(xml.documentElement, "test-suite");
        elements.forEach((element) => {
            if (this.countChildren(element) <= 1) {
                element.parentNode.removeChild(element);
            }
        });
    }
    combineTestResultsAttribute(xml1, xml2, attributeName) {
        this.addTestResultsAttribute(xml1, attributeName, this.getTestResultsAttribute(xml2, attributeName));
    }
    getTestResultsAttribute(xml, attributeName) {
        const testResults = this.collectAllElements(xml.documentElement, "test-results");
        if (testResults.length === 0) {
            return 0;
        }
        const attr = testResults[0].attributes.getNamedItem(attributeName);
        if (attr.value) {
            return Number(attr.value);
        }
        return 0;
    }
    addTestResultsAttribute(xml, attributeName, value) {
        const currentValue = this.getTestResultsAttribute(xml, attributeName);
        const testResults = this.collectAllElements(xml.documentElement, "test-results");
        const attr = testResults[0].attributes.getNamedItem(attributeName);
        if (attr) {
            attr.value = String(currentValue + value);
        }
    }
}
exports.NUnitXmlUtil = NUnitXmlUtil;
